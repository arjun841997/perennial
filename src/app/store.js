import { configureStore } from '@reduxjs/toolkit';
import { useReducer } from 'react';
import userReducer from '../features/user/redux/userSlice';

export const store = configureStore({
  reducer: {    
    user: userReducer
  },
});
