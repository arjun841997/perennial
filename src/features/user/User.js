import React, { useEffect, useState, useLayoutEffect, useRef } from 'react';
import { getusersAsync, selectUsers }from './userSlice'
import { useDispatch,  useSelector } from 'react-redux';
import styles from './User.module.css';

export function User() {
  const dispatch = useDispatch();
  const usersProps = useSelector(selectUsers);
  const [userSearch, setUserSearch] = useState("");
  const [users, setUsers] = useState([]);
  const targetRef = useRef();
  const [dimensions, setDimensions] = useState({ width:0, height: 0 });
  const [modalVisiable, setModalVisiblity] = useState(false);
  const [selectedUser, setSelectedUser] = useState({});


  useEffect(() => {
    dispatch(getusersAsync())
  },[dispatch])

  useEffect(() => {
    setUsers(usersProps)
  }, [usersProps])

  useLayoutEffect(() => {
    if (targetRef.current) {
      setDimensions({
        width: targetRef.current.offsetWidth,
        height: targetRef.current.offsetHeight
      });
    }
  }, []);

  const handleOnChangeUserSearch = (e) => {
    const searchParams = e.target.value
    setUserSearch(searchParams)    
    let filteredUsers = []    
    for(let i=0; i< usersProps.length; i++){
      const userName = usersProps[i].name.first + " " + usersProps[i].name.last            
      if(userName.toLowerCase().includes(searchParams.toLowerCase())){        
        filteredUsers.push(usersProps[i])
      }
    }    
    setUsers(filteredUsers)
  }

  return (
    <>    
    {modalVisiable && 
      <div className={styles.modal} >
        <div className={styles.modalContent}>
          <span onClick={() => {setModalVisiblity(false)}} className={styles.close}>&times;</span>
          <p><h1>{selectedUser.name.first} {selectedUser.name.last}</h1>              
          <br />     <span>ID: {selectedUser.id.value}</span><br/>
                  <span>Phone no: {selectedUser.phone}</span>
                  </p>                    
        </div>
      </div>
    }
    
    {/* <button onclick={() => { document.getElementById('myModal').style.display='block'}} class="w3-button w3-black">Open Modal</button> */}
  <div ref={targetRef}>  
    <h1>User Listing App</h1>       
    <input className={styles.input} onChange={handleOnChangeUserSearch} value={userSearch} />       
    {users && users.length > 0 && users.map((user) => {
      return (
        <div className={styles.card} onClick={() => {setModalVisiblity(true);setSelectedUser(user)}}>
            <img src={user.picture.medium} alt="user "  width="50" height="50" />
            <div className={styles.container}>            
              <h1>{user.name.first} {user.name.last}</h1>              
              {
                dimensions.width > 460 && <>
                  <span>ID: {user.id.value}</span><br/>
                  <span>Phone no: {user.phone}</span>  
                </>
              }
              <br />
              <span>Email : {user.email}</span> <br />
              <span>City : {user.location.city}</span>            
            </div>
        </div>
      )
  })}

  </div> 
  </>
  );
}
