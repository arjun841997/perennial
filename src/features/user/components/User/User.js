import React, { useEffect, useState, useLayoutEffect, useRef } from 'react';
import { getUsersAction, isLoading, selectUsers }from '../../redux/userSlice'
import { useDispatch,  useSelector } from 'react-redux';
import styles from './css/User.module.css';

export function User() {
  const dispatch = useDispatch();
  const usersProps = useSelector(selectUsers);
  const statusProps = useSelector(isLoading);  
  const [userSearch, setUserSearch] = useState("");
  const [users, setUsers] = useState([]);
  const targetRef = useRef();
  const [dimensions, setDimensions] = useState({ width:0, height: 0 });
  const [modalVisiblity, setModalVisiblity] = useState(false);
  const [selectedUser, setSelectedUser] = useState({});


  useEffect(() => {
    dispatch(getUsersAction())
  },[dispatch])

  useEffect(() => {
    setUsers(usersProps)
  }, [usersProps])

  useLayoutEffect(() => {
    if (targetRef.current) {
      setDimensions({
        width: targetRef.current.offsetWidth,
        height: targetRef.current.offsetHeight
      });
    }
  }, []);

  const handleOnChangeUserSearch = (e) => {
    const searchParams = e.target.value
    setUserSearch(searchParams)    
    let filteredUsers = []    
    for(let i=0; i< usersProps.length; i++){
      const userName = usersProps[i].name.first + " " + usersProps[i].name.last            
      if(userName.toLowerCase().includes(searchParams.toLowerCase())){        
        filteredUsers.push(usersProps[i])
      }
    }    
    setUsers(filteredUsers)
  }

  return (
    <>    
    {modalVisiblity && 
      <div className={styles.modal} >
        <div className={styles.modalContent}>
          
          <span onClick={() => {setModalVisiblity(false)}} className={styles.close}>&times;</span>
          <img src={selectedUser.picture.large} alt="user "  width="50" height="50" />
          <p><h1>{selectedUser.name.first} {selectedUser.name.last}</h1>              
          <br />     <span>ID: {selectedUser.id.value}</span><br/>
                  <span>Phone no: {selectedUser.phone}</span><br/>
                  <span>Email : {selectedUser.email}</span> <br />
                  <span>City : {selectedUser.location.city}</span>  
                  </p>                    
        </div>
      </div>
    }        
  <div ref={targetRef}>  
    <h1 className={styles.title}>User Listing App</h1>       
    <input placeholder='Search User..' className={styles.input} onChange={handleOnChangeUserSearch} value={userSearch} />       
    {statusProps === "loading" ?     
    <div className={styles.loader}></div>
    :<>
    {users && users.length > 0 ? users.map((user) => {
      return (
        <div className={dimensions.width > 460 ? styles.card : styles.cardresponsive} onClick={() => {setModalVisiblity(true);setSelectedUser(user)}}>
            <img src={user.picture.large} alt="user "  width="50" height="50" />
            <div className={styles.container}>            
              <h1>{user.name.first} {user.name.last}</h1>              
              {
                dimensions.width > 460 && <>
                  <span>ID: {user.id.value}</span><br/>
                  <span>Phone no: {user.phone}</span>  
                </>
              }
              <br />
              <span>Email : {user.email}</span> <br />
              <span>City : {user.location.city}</span>            
            </div>
        </div>
      )
     
  })
  :
  <div className={styles.alignCenter}>No data found with given search.</div>
}
  </>
}

  </div> 
  </>
  );
}
