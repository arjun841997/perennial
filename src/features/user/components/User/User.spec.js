import renderer from 'react-test-renderer';
import {User} from './User';
import { Provider } from 'react-redux';
import { store } from './../../../../app/store';

it('renders correctly', () => {
  const tree = renderer
    .create( <Provider store={store}>
      <User />
    </Provider>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});