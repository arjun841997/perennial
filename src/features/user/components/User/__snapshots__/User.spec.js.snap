// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <h1
    className="title"
  >
    User Listing App
  </h1>
  <input
    className="input"
    onChange={[Function]}
    placeholder="Search User.."
    value=""
  />
  <div
    className="alignCenter"
  >
    No data found with given search.
  </div>
</div>
`;
